clarity-ext-scripts/scripts/cancel_samples.py

```diff
@@ -0,0 +1,137 @@
+
+import sys
+import yaml
+import logging
+import time
+
+
+import click
+from retry import retry
+from requests import ConnectionError, Timeout
+
+from clarity_ext_scripts.covid.partner_api_client import PartnerAPIV7Client, ServiceRequestAlreadyExists, \
+    OrganizationReferralCodeNotFound, CouldNotCreateServiceRequest, PartnerClientAPIException, COVID_RESPONSE_FAILED, \
+    MoreThanOneOrganizationReferralCodeFound
+
+log = logging.getLogger(__name__)
+logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)
+
+
+@click.group()
+@click.option("--config")
+@click.pass_context
+def cli(ctx, config):
+    with open(config) as f:
+        client_config = yaml.safe_load(f)
+    client = PartnerAPIV7Client(**client_config)
+    ctx.obj['CLIENT'] = client
+
+
+@cli.command()
+@click.argument("filename")
+@click.pass_context
+def validate(ctx, filename):
+    log.info("Validating...")
+    client = ctx.obj["CLIENT"]
+
+    barcodes = read_barcodes_from_file(filename)
+    log.info("Found {} barcodes in file.".format(len(barcodes)))
+
+    service_request_ids = []
+    more_than_one_ref_found = []
+    # Check if there already exists a service request
+    for barcode in barcodes:
+        try:
+            service_request_id = search_for_service_request(client, barcode)
+            service_request_ids.append(service_request_id)
+        except OrganizationReferralCodeNotFound as e:
+            pass
+        except MoreThanOneOrganizationReferralCodeFound as e:
+            more_than_one_ref_found.append(barcode)
+
+    log.info("Found service requests for for {}/{} of the barcodes.".format(
+        len(service_request_ids), len(barcodes)))
+    log.info("There were multiple service requests for {} barcodes.".format(
+        len(more_than_one_ref_found)))
+
+
+# Since it will cause us a headache is we fail half way through reporting
+# due to connection problems, retry this.
+@retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+def send_results_to_partner(client, service_request_id):
+    client.post_diagnosis_report(service_request_id=service_request_id,
+                                 diagnosis_result=COVID_RESPONSE_FAILED,
+                                 analysis_results=[{"value": -1}])
+
+
+@retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+def create_anonymous_service_request(client, barcode):
+    return client.create_anonymous_service_request(barcode)
+
+
+@retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+def search_for_service_request(client, barcode):
+    search_found = client.search_for_service_request(
+        org="http://uri.d-t.se/id/Identifier/i-referral-code", org_referral_code=barcode)
+    service_request_id = search_found["resource"]["id"]
+    return service_request_id
+
+
+@cli.command()
+@click.argument("filename")
+@click.pass_context
+def execute(ctx, filename):
+    log.info("Executing...")
```

Comment:
I wouldn't say no to a comment indicating what the method does, since the name is quite generic.